File: apps/desktop/src/domain/media.ts
import type { Attachment, AttachmentPreview } from "../models";

export type AttachmentRenderKind = "image" | "audio" | "video" | "document" | "file";

export function attachmentRenderKind(
  attachment: Attachment,
  preview: AttachmentPreview | null,
): AttachmentRenderKind {
  if (preview?.mediaType.startsWith("image/")) {
    return "image";
  }
  if (preview?.mediaType.startsWith("audio/")) {
    return "audio";
  }
  if (preview?.mediaType.startsWith("video/")) {
    return "video";
  }
  if (preview?.mediaType === "application/pdf") {
    return "document";
  }

  switch (attachment.kind) {
    case "audio":
      return "audio";
    case "gif":
    case "photo":
    case "sticker":
      return "image";
    case "video":
      return "video";
    default:
      return "file";
  }
}

export function canRequestAttachmentPreview(attachment: Attachment): boolean {
  return ["audio", "gif", "photo", "sticker", "video", "unknown"].includes(attachment.kind);
}

File: crates/whatsvault-core/tests/media.rs
use whatsvault_core::{
    media::{
        attachment_display_label, attachment_kind_from_filename,
        attachment_kind_from_mime_or_filename,
    },
    AttachmentKind,
};

#[test]
fn classifies_attachment_kind_from_filename_extensions() {
    assert_eq!(
        attachment_kind_from_filename("photo.JPG"),
        AttachmentKind::Photo
    );
    assert_eq!(
        attachment_kind_from_filename("voice.opus"),
        AttachmentKind::Audio
    );
    assert_eq!(
        attachment_kind_from_filename("clip.mp4"),
        AttachmentKind::Video
    );
    assert_eq!(
        attachment_kind_from_filename("animated.gif"),
        AttachmentKind::Gif
    );
    assert_eq!(
        attachment_kind_from_filename("document.pdf"),
        AttachmentKind::Unknown
    );
}

#[test]
fn prefers_mime_type_when_it_is_more_specific_than_filename() {
    assert_eq!(
        attachment_kind_from_mime_or_filename(Some("audio/ogg"), "media.bin"),
        AttachmentKind::Audio
    );
    assert_eq!(
        attachment_kind_from_mime_or_filename(Some("image/jpeg"), "media.bin"),
        AttachmentKind::Photo
    );
    assert_eq!(
        attachment_kind_from_mime_or_filename(Some("video/mp4"), "media.bin"),
        AttachmentKind::Video
    );
    assert_eq!(
        attachment_kind_from_mime_or_filename(None, "sticker.webp"),
        AttachmentKind::Photo
    );
}

#[test]
fn exposes_human_attachment_labels_for_chat_previews() {
    assert_eq!(attachment_display_label(AttachmentKind::Photo), "Photo");
    assert_eq!(
        attachment_display_label(AttachmentKind::Audio),
        "Voice message"
    );
    assert_eq!(
        attachment_display_label(AttachmentKind::Unknown),
        "Attachment"
    );
}
